### PR TITLE
RFR: Async query of mistral results.

### PR DESCRIPTION
**Pull request by lakshmi-kannan**
## TODO
- [X] - Create a boiler plate Querier class which makes query on eventlets.
- [X] - Provide an interface for implementing query
- [X] - Save the query results in ActionExecutionDB
- [X] - Data models for persisting action execution states. 
- [X] - Provide a mechanism to bootstrap the querier with query contexts after restart
- [X] - Provide a mechanism to add a query context in a live system [wiring]
- [X] - Fix existing tests.
- [X] - Add tests - [Continuing]
- [X] - Test one box deployment
- [X] - Rename things
- [X] - Refactor runner interface
- [X] - Get rid of mistral callback https://github.com/StackStorm/st2mistral/pull/5
- [X] - Update mistral examples to not use callbacks
- [X] - Update docs (mistral)
- [X] - Packaging changes
- [X] - Resolve merge conflicts
- [X] - Changelog

## Sample output
```
(virtualenv)~/stanley git:mistral_get_results ❯❯❯ st2 execution get 54a3559a0640fd5c5ffefebc                           ✭ ✱ ◼
STATUS: succeeded
RESULT:
{
    "tasks": [
        {
            "wf_name": "examples.mistral-basic.demo",
            "name": "callback-on-success",
            "created_at": "2014-12-31 01:47:06",
            "updated_at": "2014-12-31 01:47:07",
            "state": "SUCCESS",
            "result": "null",
            "input": "{"state": "SUCCESS", "result": "foo\n"}",
            "output": "{"task": {"callback-on-success": null}}",
            "id": "35fc5388-e4b9-4687-8f9a-df54b6abfd98",
            "execution_id": "54158197-eba3-46d3-8017-e5dc46fc4680"
        },
        {
            "wf_name": "examples.mistral-basic.demo",
            "name": "run-cmd",
            "created_at": "2014-12-31 01:47:06",
            "updated_at": "2014-12-31 01:47:06",
            "state": "SUCCESS",
            "result": "{"stderr": "", "stdout": "foo\n"}",
            "input": "{"ref": "core.local", "parameters": {"cmd": "echo foo"}}",
            "output": "{"task": {"run-cmd": {"stderr": "", "stdout": "foo\n"}}, "stderr": "", "stdout": "foo\n"}",
            "id": "dfc664c0-fd6a-426d-9f6c-09de13e4aada",
            "execution_id": "54158197-eba3-46d3-8017-e5dc46fc4680"
        }
    ]
}
(virtualenv)~/stanley git:mistral_get_results ❯❯❯
```

### CODE REVIEW

**Comment by m4dcoder** [comment]
I have a few opinions. 1) Overall, the solution here is a bit sophisticated.  Just reiterating what we talked about last week.  I was not expecting something as complicated and I question if there's a simpler solution.  Or maybe I'm oversimplifying.  2) How come we are not using the context field in ActionExecution but create a new ActionExecutionState model?  3) What does the AsynActionRunner abstraction adds?  The abstract class does not add anything new on top of the ActionRunner.  Aside from these design related questions, the implementation looks fine.  I don't want to block.  So if this is tested, achieve the expected result, and you're convinced of the current design, please feel free to merge.

File: st2actions/st2actions/cmd/st2resultstracker.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+
+LOG = logging.getLogger(__name__)
+
+
+eventlet.monkey_patch(
+    os=True,
+    select=True,
+    socket=True,
+    thread=False if '--use-debugger' in sys.argv else True,
```
Comment (manasdk) [inline st2actions/st2actions/cmd/st2resultstracker.py]: I don't think we really use this but if you care you user a debugger keep it around.

File: st2actions/st2actions/resultstracker.py [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+
+        for querier, contexts in six.iteritems(query_contexts_dict):
+            LOG.info('Found %d pending actions for query module %s', len(contexts), querier)
+            querier.add_queries(query_contexts=contexts)
+
+    def get_querier(self, query_module_name):
+        if (query_module_name not in self._queriers and
+                query_module_name not in self._failed_imports):
+            try:
+                querier = self._import_query_module(query_module_name)
```
Comment (manasdk) [inline st2actions/st2actions/resultstracker.py]: suggestion -

querier -> query_module
instance -> querier

self._queriers contains the instance therefore the variable names will end up aligning.

File: st2actions/st2actions/query/base.py [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+                eventlet.greenthread.sleep(self._no_workers_sleep_time)
+            self._fire_queries()
+
+    def add_queries(self, query_contexts=[]):
+        LOG.debug('Adding queries to querier: %s' % query_contexts)
+        for query_context in query_contexts:
+            self._query_contexts.put((time.time(), query_context))
+
+    def is_started(self):
+        return self._is_started
```
Comment (manasdk) [inline st2actions/st2actions/query/base.py]: _started?

File: st2actions/st2actions/query/base.py [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+        self.query_context = query_context
+        self.query_module = query_module
+
+    @classmethod
+    def from_model(cls, model):
+        return QueryContext(str(model.id), str(model.execution_id), model.query_context,
+                            model.query_module)
+
+    def __repr__(self):
+        return ('<QueryContext id=%s,execution_id=%s,query_context=%s>' %
```
Comment (manasdk) [inline st2actions/st2actions/query/base.py]: would be nice if repr implementation would be eval(repr(c)) == c. Anyway, I guess as long as this is unambigious.

File: st2common/st2common/models/api/action.py [hunk 5]
```diff
@@ -273,3 +282,51 @@ def to_model(cls, execution):
...
+                "type": "string",
+                "description": "Name of the query module.",
+                "required": True
+            }
+        },
+        "additionalProperties": False
+    }
+
+    @classmethod
+    def from_model(cls, model):
```
Comment (manasdk) [inline st2common/st2common/models/api/action.py]: The super class impl of this method is pretty much the same. Can you use that instead?

**Comment by manasdk** [comment]
LGTM.

File: st2actions/st2actions/cmd/st2resultstracker.py (hunk 1 above)
Comment (lakshmi-kannan) [inline st2actions/st2actions/cmd/st2resultstracker.py]: Let's have this consistent with other monkey patching.

**Comment by lakshmi-kannan** [comment]
> I have a few opinions. 1) Overall, the solution here is a bit sophisticated. Just reiterating what we talked about last week. I was not expecting something as complicated and I question if there's a simpler solution. Or maybe I'm oversimplifying. 2) How come we are not using the context field in ActionExecution but create a new ActionExecutionState model? 3) What does the AsynActionRunner abstraction adds? The abstract class does not add anything new on top of the ActionRunner. Aside from these design related questions, the implementation looks fine. I don't want to block. So if this is tested, achieve the expected result, and you're convinced of the current design, please feel free to merge.

@m4dcoder 1) We discussed it already :). 2) The problem is bootstrapping. When st2 shuts down and restarts, it needs to fetch all those execution objects that have context that are query worthy. A separate state object makes this easier. Also, state objects are short lived. They are deleted as soon as we get the results. Having them in context would mean we'll keep things forever or clear context in execution object. This is tricky. 3) AsyncActionRunner run returns a tuple of (status, results, query_context). ActionRunner run returns a tuple (status, results). My plan was to add type annotations so we can validate the return types. It turns out this is hard in python 2.7 but possible in python 3. See https://github.com/StackStorm/st2/pull/929/files#diff-ef7b3e3b028fe99ecaed2faaad32fd11R122. There I check the type of runner and catch the tuple appropriately. This was the thing I brought up on slack and asked for suggestions. I am happy to change this if there are better ways. I haven't heard any yet.

File: st2actions/st2actions/query/base.py (hunk 4 above)
Comment (lakshmi-kannan) [inline st2actions/st2actions/query/base.py]: I used this from one of our existing reprs. If you point me to something you like, I can fix this :)

File: st2actions/st2actions/query/base.py (hunk 3 above)
Comment (lakshmi-kannan) [inline st2actions/st2actions/query/base.py]: oops.

Comment (lakshmi-kannan) [inline st2actions/st2actions/query/base.py]: Fixed d1d80a4

File: st2actions/st2actions/resultstracker.py (hunk 2 above)
Comment (lakshmi-kannan) [inline st2actions/st2actions/resultstracker.py]: Fixed 999e6c7

File: st2common/st2common/models/api/action.py (hunk 5 above)
Comment (lakshmi-kannan) [inline st2common/st2common/models/api/action.py]: That's what I am using in line below.

**Pull request merged**